src/experimental/coroutines.rs

```diff
@@ -46,51 +54,83 @@ impl CoroutineState {
 }
 
 pub(crate) struct CoroutinesContext {
-    coroutines: Vec<CoroutineState>,
+    coroutines: GenerationalStorage<CoroutineState>,
     active_coroutine_now: Option<f64>,
     active_coroutine_delta: Option<f64>,
 }
 
 impl CoroutinesContext {
     pub fn new() -> CoroutinesContext {
         CoroutinesContext {
-            coroutines: Vec::with_capacity(1000),
+            coroutines: GenerationalStorage::new(),
             active_coroutine_now: None,
             active_coroutine_delta: None,
         }
     }
 
     pub fn update(&mut self) {
-        for future in &mut self.coroutines {
-            if let CoroutineState::Running(f) = future {
+        self.coroutines.retain(|coroutine| {
+            if let CoroutineState::Running(ref mut f) = coroutine {
                 if f.manual_poll == false {
                     if let Some(v) = resume(&mut f.future) {
-                        *future = CoroutineState::Value(v);
+                        if f.has_value {
+                            *coroutine = CoroutineState::Value(v);
+                        } else {
+                            return false;
+                        }
                     }
                 }
             }
-        }
+
+            true
+        });
+    }
+
+    pub(crate) fn allocated_memory(&self) -> usize {
+        self.coroutines.allocated_memory()
+    }
+
+    pub(crate) fn active_coroutines_count(&self) -> usize {
+        self.coroutines.count()
     }
+
 }
 #[derive(Clone, Copy, Debug)]
 pub struct Coroutine<T = ()> {
-    id: usize,
+    id: GenerationalId,
     _phantom: PhantomData<T>,
 }
 
-impl<T: 'static> Coroutine<T> {
+impl<T: 'static + Any> Coroutine<T> {
     /// Returns true if the coroutine finished or was stopped.
     pub fn is_done(&self) -> bool {
         let context = &get_context().coroutines_context;
 
-        context.coroutines[self.id].is_value() || context.coroutines[self.id].is_nothing()
+        let coroutine = context.coroutines.get(self.id);
+
+        if let Some(coroutine) = coroutine {
+            return coroutine.is_value() || coroutine.is_nothing();
+        }
+
+        return true;
     }
 
-    pub fn retrieve(self) -> Option<T> {
+    pub fn retrieve(&self) -> Option<T> {
         let context = &mut get_context().coroutines_context;
 
-        if let Some(v) = context.coroutines[self.id].take_value() {
-            return Some(*v.downcast().unwrap());
+        // () is a special case. Futures with () as a return type do not keep
```

Comment:
This special casing shouldn't be necessary? a `Coroutine<()>` will still produce a `Box<()> as Box<dyn Any>` so the general case should work fine. Except that this special case will always return Some(()), even multiple times, is that intentional?

Reply:
nevermind I see the reason now